### ISSUE TITLE
[BUG] Incorrect speaker mapping and driver setting on Inspiron 16 Plus 7620

### ISSUE THREAD

**Issue by lakotamm** [thesofproject/linux#3825]
There are 4 speakers on Inspiron 16 Plus 7620. 2 top firing (high frequency) and 2 bottom firing (lower frequency). By default only the bottom firing speakers are detected by the sof-hda-dsp driver. 
 
According to alsa-info the laptop is using ALC3254 codec and only speakers on pin 0x14 are detected. Using the HDA-Intel driver we managed to find the correct connections for the second pair of speakers - 0x17 as well as correct assignment of amplifiers. Here is a kernel patch (that has not been proposed as a change yet) which makes all 4 speakers work on HDA-Intel driver on kernel 5.20 RC0:
[0001-ALSA-patch_realtek-Fix-Dell-Inspiron-Plus-16.patch.txt](https://github.com/thesofproject/sof/files/9351028/0001-ALSA-patch_realtek-Fix-Dell-Inspiron-Plus-16.patch.txt)

There is still one issue left - the top speakers start playing sound only after ~0,5s of silence. But in real life it is almost not noticeable since the bottom firing speakers are already playing sound.

Unfortunately the HDA-Intel driver does not work with the inbuilt microphone (and I cannot find a way to fix it by remapping pins) so it does not really seem like a good option for this laptop in the first place.

Therefore my question is - **Would it be possible to implement a similar fix to the sof-hda-dsp driver?**
 
- Tested distributions: Fedora 36, Ubuntu 22.04, Arch
- Kernel: 5.18, 5.19, 5.20RC0
- alsa-sof-firmware /sof-firmware 2.1.1

Alsa-info with sof-hda-dsp driver:
http://alsa-project.org/db/?f=ab7af01dfa223d62dc6ec0b398c87f163fa5b50c

Alsa info with HDA-Intel driver with the kernel patch applied:
http://alsa-project.org/db/?f=6df45ca8fff2405febb97be8b314c24528417364

Thank you very much for your help!

**Comment by lgirdwood** [comment]
@plbossart @ujfalusi question for you guys.

**Comment by plbossart** [comment]
@lakotamm in theory the settings for HDaudio are identical for the legacy snd-hda-intel and the SOF drivers, we use the same code. It's not clear to me why the behavior would be different between the two cases.
To be clearer, there is nothing in the SOF driver related to the HDaudio settings, so if there is an issue it's a problem related to the interface or sequence between SOF and legacy driver.

Adding @kv2019i for comments.

**Comment by lakotamm** [comment]
You are right. I am sorry for the confusion. **The speakers are actually working fine when the patch is applied to the 6.0RC1 kernel and the sof-hda-dsp driver is loaded**. For some reason I thought that I need to be running the snd-hda-intel driver in order for the HDaudio to be running and I did not even test the patch with the sof-hda-dsp driver. Oh well...

Thank you for the clarification!

So I guess that the next step is:
- to propose the patch for the kernel 6.0.
- sometimes in the future deal with headset microphone not outputing any sound. But that is another (way less important) topic.

If nobody else has comments, this can be closed.

**Comment by NGStaph** [comment]
I'm facing the exact same error on Inspiron 16 Plus 7360 and haven't been able to use the patch linked above, [or this one](https://github.com/tiwai/sound/commit/2912cdda734d9136615ed05636d9fcbca2a7a3c5.patch). I'm new to this and would like a little bit of advice on how I ought to proceed with this.

**Comment by plbossart** [comment]
@NGStaph this is an 1.5 yr old issue, the patch is already in recent kernels. Updating your kernel should work with zero work required on your side. If not, file a new issue.

**Comment by lakotamm** [comment]
> I'm facing the exact same error on Inspiron 16 Plus 7360 and haven't been able to use the patch linked above, [or this one](https://github.com/tiwai/sound/commit/2912cdda734d9136615ed05636d9fcbca2a7a3c5.patch). I'm new to this and would like a little bit of advice on how I ought to proceed with this.

Yeah, 7360 likely has a different ID, which probably just needs to be added to the kernel in this function:

```
" static const struct snd_pci_quirk alc269_fixup_tbl[] = {
@@ -9195,6 +9230,8 @@ static const struct snd_pci_quirk alc269_fixup_tbl[] = {
 	SND_PCI_QUIRK(0x1028, 0x0a9e, "Dell Latitude 5430", ALC269_FIXUP_DELL4_MIC_NO_PRESENCE),
 	SND_PCI_QUIRK(0x1028, 0x0b19, "Dell XPS 15 9520", ALC289_FIXUP_DUAL_SPK),
 	SND_PCI_QUIRK(0x1028, 0x0b1a, "Dell Precision 5570", ALC289_FIXUP_DUAL_SPK),
+	SND_PCI_QUIRK(0x1028, 0x0b37, "Dell Inspiron 16 Plus 7620 2-in-1", ALC295_FIXUP_DELL_INSPIRON_TOP_SPEAKERS),
+	SND_PCI_QUIRK(0x1028, 0x0b71, "Dell Inspiron 16 Plus 7620", ALC295_FIXUP_DELL_INSPIRON_TOP_SPEAKERS),"
```

As you can see, right now the patch gets applied only to 7620 and 7620 2 in 1. It would be worth getting the ID of 7360 and applying the same SND_PCI_QUIRK to it (which literally just means adding one line of code). And then recompiling the kernel with the new patch.

Link to the patch which was submitted to mainstream:
https://github.com/lakotamm/Dell-Inspiron-16-Plus-7620-Linux-Config/blob/main/Kernel_patch_to_fix_speakers_on_Inspiron_16_Plus_7620.patch

**Comment by NGStaph** [comment]
Thanks so much. How would i go about finding the ID? I'd appreciate it if you could help with a couple of questions. In case it helps here's my [alsa-info](http://alsa-project.org/db/?f=4a4fcd6f57ff8af0291e10640883ca9d988ef93c)

i think the realtek codec has changed? Does that imply that the pin-layout may have changed as well? Will i have to find the pins using hdajackretask? Or are the reverse engineering links in the [suggestions before filing a bug](https://thesofproject.github.io/latest/getting_started/intel_debug/suggestions.html) section a safer bet considering i'm having trouble getting hdajackretask to work?

**Comment by lakotamm** [comment]
> Thanks so much. How would i go about finding the ID? I'd appreciate it if you could help with a couple of questions. In case it helps here's my [alsa-info](http://alsa-project.org/db/?f=4a4fcd6f57ff8af0291e10640883ca9d988ef93c)
> 
> i think the realtek codec has changed? Does that imply that the pin-layout may have changed as well? Will i have to find the pins using hdajackretask? Or are the reverse engineering links in the [suggestions before filing a bug](https://thesofproject.github.io/latest/getting_started/intel_debug/suggestions.html) section a safer bet considering i'm having trouble getting hdajackretask to work?


From your alsa info:
```
Subsystem: Dell Device [1028:0c28]
..
Subsystem Id: 0x10280c28
```

As I can see, the controller is still the same:
`Card sysdefault:1 'sofhdadsp'/'DellInc.-Inspiron16Plus7630--0HRN2X'
  Mixer name	: 'Realtek ALC3254'`
`Card sysdefault:1 'sofhdadsp'/'DellInc.-Inspiron16Plus7620--0V98G2'
  Mixer name	: 'Realtek ALC3254'`
  
So hopefully you just need to add:
`SND_PCI_QUIRK(0x1028, 0x0c28, "Dell Inspiron 16 Plus 7360", ALC295_FIXUP_DELL_INSPIRON_TOP_SPEAKERS),`

In this file:
https://github.com/torvalds/linux/blob/master/sound/pci/hda/patch_realtek.c
In this function:
`static const struct snd_pci_quirk alc269_fixup_tbl[] = {`
After my 7260:
```
	SND_PCI_QUIRK(0x1028, 0x0b37, "Dell Inspiron 16 Plus 7620 2-in-1", ALC295_FIXUP_DELL_INSPIRON_TOP_SPEAKERS),
	SND_PCI_QUIRK(0x1028, 0x0b71, "Dell Inspiron 16 Plus 7620", ALC295_FIXUP_DELL_INSPIRON_TOP_SPEAKERS),
```

When I was using hdajackretask to change the pins, the laptop kept misbehaving really badly. So I would really recommend trying to simply create a kernel patch, add the exactly same Quirk to your new ID (which that single line should do) and recompile the kernel. Otherwise Philipp Jungkamp <p.jungkamp@gmx.net> - helped me with debugging the whole thing and creating a patch.

**Comment by NGStaph** [comment]
Cheers! I got it to work!

thanks a lot for your help.
